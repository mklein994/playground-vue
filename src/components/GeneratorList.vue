<script setup lang="ts">
import { ref, watchEffect } from "vue";

const props = withDefaults(
  defineProps<{
    initialMax?: number;
  }>(),
  {
    initialMax: 25,
  },
);

function* iteratorGenerator(max: number) {
  let i = 1;
  while (i <= max) {
    yield i++;
  }
}

const fizzBuzz = (max: number) => {
  const list = [];
  for (let i of iteratorGenerator(max)) {
    list.push(
      i % 15 === 0
        ? "FizzBuzz"
        : i % 3 === 0
          ? "Fizz"
          : i % 5 === 0
            ? "Buzz"
            : i,
    );
  }
  return list;
};

const max = ref(props.initialMax);

const list = ref<(string | number)[]>([]);

watchEffect(() => {
  list.value = fizzBuzz(max.value);
});
</script>

<template>
  <div class="generator-list">
    <label for="max">Max:</label>
    <input
      id="max"
      v-model="max"
      type="number"
      name="max"
      min="1"
      class="tw-form-input"
    />

    <ol>
      <li v-for="(value, index) of list" :key="`${value}${index}`">
        <span :class="{ 'is-number': Number.isInteger(value) }">{{
          value
        }}</span>
      </li>
    </ol>
  </div>
</template>

<style scoped>
.is-number {
  color: dodgerblue;
}
</style>
