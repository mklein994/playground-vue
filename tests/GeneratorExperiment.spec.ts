import { flushPromises, mount } from "@vue/test-utils";
import { describe, expect, it } from "vitest";
import { defineComponent, h, Suspense } from "vue";

import GeneratorList from "@/components/GeneratorList.vue";

describe.concurrent("GeneratorExperiment", () => {
  const testComponentBuilder = (max?: number) =>
    defineComponent(function TestComponent() {
      return () =>
        h(Suspense, () =>
          h(GeneratorList, max == null ? undefined : { initialMax: max }),
        );
    });

  describe.concurrent("all in order", async () => {
    const wrapper = mount(testComponentBuilder());

    await flushPromises();

    it.each([
      [1, "1"],
      [2, "2"],
      [3, "Fizz"],
      [4, "4"],
      [5, "Buzz"],
      [6, "Fizz"],
      [7, "7"],
      [8, "8"],
      [9, "Fizz"],
      [10, "Buzz"],
      [11, "11"],
      [12, "Fizz"],
      [13, "13"],
      [14, "14"],
      [15, "FizzBuzz"],
      [16, "16"],
    ])("fizzbuzz: %i -> %s", (index, expected) => {
      expect(wrapper.get(`li:nth-child(${index})`).text()).toBe(expected);
    });
  });

  it("matches snapshot", async () => {
    const wrapper = mount(testComponentBuilder());

    await flushPromises();

    expect(wrapper.html()).toMatchSnapshot();
  });

  describe.concurrent("max", () => {
    it("handles a different initial value", async () => {
      const wrapper = mount(testComponentBuilder(10));

      await flushPromises();

      expect(wrapper.get("li:last-child").text()).toBe("Buzz");
    });

    it("updates when changed", async () => {
      const wrapper = mount(testComponentBuilder(10));

      await flushPromises();

      const last = () => wrapper.get("li:last-child");
      expect(wrapper.findAll("li")).toHaveLength(10);
      expect(last().text()).toBe("Buzz");

      await wrapper.get("#max").setValue(3);

      await flushPromises();

      expect(wrapper.findAll("li")).toHaveLength(3);
      expect(last().text()).toBe("Fizz");
    });
  });
});
