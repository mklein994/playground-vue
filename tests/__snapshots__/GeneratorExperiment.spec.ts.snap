// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`GeneratorExperiment > matches snapshot 1`] = `
"<div data-v-caa99902="" class="generator-list"><label data-v-caa99902="" for="max">Max:</label><input data-v-caa99902="" id="max" type="number" name="max" min="1" class="tw-form-input">
  <ol data-v-caa99902="">
    <li data-v-caa99902=""><span data-v-caa99902="" class="is-number">1</span></li>
    <li data-v-caa99902=""><span data-v-caa99902="" class="is-number">2</span></li>
    <li data-v-caa99902=""><span data-v-caa99902="" class="">Fizz</span></li>
    <li data-v-caa99902=""><span data-v-caa99902="" class="is-number">4</span></li>
    <li data-v-caa99902=""><span data-v-caa99902="" class="">Buzz</span></li>
    <li data-v-caa99902=""><span data-v-caa99902="" class="">Fizz</span></li>
    <li data-v-caa99902=""><span data-v-caa99902="" class="is-number">7</span></li>
    <li data-v-caa99902=""><span data-v-caa99902="" class="is-number">8</span></li>
    <li data-v-caa99902=""><span data-v-caa99902="" class="">Fizz</span></li>
    <li data-v-caa99902=""><span data-v-caa99902="" class="">Buzz</span></li>
    <li data-v-caa99902=""><span data-v-caa99902="" class="is-number">11</span></li>
    <li data-v-caa99902=""><span data-v-caa99902="" class="">Fizz</span></li>
    <li data-v-caa99902=""><span data-v-caa99902="" class="is-number">13</span></li>
    <li data-v-caa99902=""><span data-v-caa99902="" class="is-number">14</span></li>
    <li data-v-caa99902=""><span data-v-caa99902="" class="">FizzBuzz</span></li>
    <li data-v-caa99902=""><span data-v-caa99902="" class="is-number">16</span></li>
    <li data-v-caa99902=""><span data-v-caa99902="" class="is-number">17</span></li>
    <li data-v-caa99902=""><span data-v-caa99902="" class="">Fizz</span></li>
    <li data-v-caa99902=""><span data-v-caa99902="" class="is-number">19</span></li>
    <li data-v-caa99902=""><span data-v-caa99902="" class="">Buzz</span></li>
    <li data-v-caa99902=""><span data-v-caa99902="" class="">Fizz</span></li>
    <li data-v-caa99902=""><span data-v-caa99902="" class="is-number">22</span></li>
    <li data-v-caa99902=""><span data-v-caa99902="" class="is-number">23</span></li>
    <li data-v-caa99902=""><span data-v-caa99902="" class="">Fizz</span></li>
    <li data-v-caa99902=""><span data-v-caa99902="" class="">Buzz</span></li>
  </ol>
</div>"
`;
